optuna/visualization/_pareto_front.py

```diff
@@ -133,42 +153,89 @@ def _get_pareto_front_2d(
                 "lower than 0."
             )
 
-    data = [
-        go.Scatter(
-            x=[t.values[axis_order[0]] for t in trials[n_best_trials:]],
-            y=[t.values[axis_order[1]] for t in trials[n_best_trials:]],
-            text=[_make_hovertext(t) for t in trials[n_best_trials:]],
-            mode="markers",
-            hovertemplate="%{text}<extra>Trial</extra>",
-            marker={
-                "line": {"width": 0.5, "color": "Grey"},
-                "color": [t.number for t in trials[n_best_trials:]],
-                "colorscale": "Blues",
-                "colorbar": {
-                    "title": "#Trials",
-                },
-            },
-            showlegend=False,
-        ),
-        go.Scatter(
-            x=[t.values[axis_order[0]] for t in trials[:n_best_trials]],
-            y=[t.values[axis_order[1]] for t in trials[:n_best_trials]],
-            text=[_make_hovertext(t) for t in trials[:n_best_trials]],
-            mode="markers",
-            hovertemplate="%{text}<extra>Best Trial</extra>",
-            marker={
-                "line": {"width": 0.5, "color": "Grey"},
-                "color": [t.number for t in trials[:n_best_trials]],
-                "colorscale": "Reds",
-                "colorbar": {
-                    "title": "#Best trials",
-                    "x": 1.1 if include_dominated_trials else 1,
-                    "xpad": 40,
-                },
-            },
-            showlegend=False,
-        ),
-    ]
+    if constraints_func is None:
+        data = [
+            go.Scatter(
+                x=[t.values[axis_order[0]] for t in trials[n_best_trials:]],
+                y=[t.values[axis_order[1]] for t in trials[n_best_trials:]],
+                text=[_make_hovertext(t) for t in trials[n_best_trials:]],
+                mode="markers",
+                hovertemplate="%{text}<extra>Trial</extra>",
+                marker=_make_marker(
+                    trials[n_best_trials:],
+                    include_dominated_trials,
+                    use_constraints_func=False,
+                    dominated_samples=True,
+                ),
+            ),
+            go.Scatter(
+                x=[t.values[axis_order[0]] for t in trials[:n_best_trials]],
+                y=[t.values[axis_order[1]] for t in trials[:n_best_trials]],
+                text=[_make_hovertext(t) for t in trials[:n_best_trials]],
+                mode="markers",
+                hovertemplate="%{text}<extra>Best Trial</extra>",
+                marker=_make_marker(
+                    trials[:n_best_trials],
+                    include_dominated_trials,
+                    use_constraints_func=False,
+                    dominated_samples=False,
+                ),
+            ),
+        ]
```

Comment:
I'm just wondering if we can reduce lines here by iterating over pairs of range and label.
Just like:

```python
        data = [
            go.Scatter(
                x=[t.values[axis_order[0]] for t in trials[slice]],
                y=[t.values[axis_order[1]] for t in trials[slice]],
                text=[_make_hovertext(t) for t in trials[slice]],
                mode="markers",
                hovertemplate="%{text}<extra>" label "</extra>",
                marker=_make_marker(
                    trials[:n_best_trials],
                    include_dominated_trials,
                    use_constraints_func=False,
                    dominated_samples=False,
                ),
            )
            for slice, label in ((n_best_trials:, "Trial"), (:n_best_trials, "Best Trial"))
        ]
```

(I don't check if the code actually works mm)

Reply:
I found this is addressed in https://github.com/optuna/optuna/pull/3128, please ignore.